src/Language/PureScript/TypeChecker/Types.hs

```diff
@@ -357,7 +357,13 @@ infer' (Abs binder ret)
       ty <- freshType
       withBindingGroupVisible $ bindLocalVariables [(arg, ty, Defined)] $ do
         body@(TypedValue _ _ bodyTy) <- infer' ret
-        return $ TypedValue True (Abs (VarBinder arg) body) $ function ty bodyTy
+        let (qs, cs, ty') = moveQuantifiersToFrontOverFunction (function ty bodyTy)
+            expr = Abs (VarBinder arg) body
+        -- if we shifted some quantifiers/constraints then we'll need to check against the new type
+        expr' <- if null qs && null cs
+          then return expr
+          else check expr ty'
```

Comment:
Can you explain why we need to check again here? It definitely seems odd to me to have to traverse the same terms twice.